crates/ide_assists/src/handlers/lambdify_for_each.rs

```diff
@@ -0,0 +1,388 @@
+use stdx::format_to;
+use hir::HirDisplay;
+use syntax::{
+    SyntaxKind,
+    ast::{self, edit_in_place::Indent, HasArgList, Pat, Expr, BinaryOp, ArithOp},
+    AstNode,
+};
+use ide_db::helpers::node_ext::walk_pat;
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: convert_if_to_filter
+//
+// Converts a if into a filter when placed in a for_each().
+// ```
+// # //- minicore: iterators
+// # use core::iter;
+// fn main() {
+//     let it = core::iter::repeat(92);
+//     it.for_each$0(|x| {
+//         if x > 4 {
+//             println!("{}", x);
+//         };
+//     });
+// }
+// ```
+// ->
+// ```
+// # use core::iter;
+// fn main() {
+//     let it = core::iter::repeat(92);
+//     it.filter(|&x| x > 4).for_each(|x| {
+//         println!("{}", x);
+//     });
+// }
+// ```
+pub(crate) fn convert_if_to_filter(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let method = ctx.find_node_at_offset::<ast::MethodCallExpr>()?;
+
+    let closure = match method.arg_list()?.args().next()? {
+        ast::Expr::ClosureExpr(expr) => expr,
+        _ => return None,
+    };
+
+    let (method, receiver) = validate_method_call_expr(ctx, method)?;
+
+    let param_list = closure.param_list()?;
+    let param = param_list.params().next()?.pat()?;
+    let body = closure.body()?;
+
+    let range = method.syntax().text_range();
+
+    let if_expr = match body.clone() {
+        Expr::IfExpr(if_expr) => {
+            if_expr
+        },
+        Expr::BlockExpr(block) => {
+            let mut stmts = block.statements();
+            let fst_stmt = stmts.next()?;
+            continue_iff(stmts.next().is_none())?; // Only one statement
```

Comment:
This assist doesn't seem too useful if it only applies when block consists of a single if expression